internal/api/impl/auth/oidc.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2023 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/gorilla/securecookie"
+	"github.com/labstack/echo/v4"
+	"github.com/perses/perses/internal/api/shared"
+	"github.com/perses/perses/internal/api/shared/crypto"
+	"github.com/perses/perses/internal/api/shared/route"
+	"github.com/perses/perses/pkg/model/api/config"
+	"github.com/sirupsen/logrus"
+	"github.com/zitadel/oidc/v3/pkg/client/rp"
+	httphelper "github.com/zitadel/oidc/v3/pkg/http"
+	"github.com/zitadel/oidc/v3/pkg/oidc"
+)
+
+type oIDCEndpoint struct {
+	relyingParty    rp.RelyingParty
+	jwt             crypto.JWT
+	tokenManagement tokenManagement
+	slugID          string
+	urlParams       map[string]string
+}
+
+func newOIDCEndpoint(provider config.OIDCProvider, jwt crypto.JWT) (authEndpoint, error) {
+	// As the cookie is used only at login time, we don't need a persistent value here.
+	key := securecookie.GenerateRandomKey(16)
+	cookieHandler := httphelper.NewCookieHandler(key, key)
+	client := &http.Client{
+		Timeout: time.Minute,
+	}
+	options := []rp.Option{
+		rp.WithCookieHandler(cookieHandler),
+		rp.WithVerifierOpts(rp.WithIssuedAtOffset(5 * time.Second)),
+		rp.WithHTTPClient(client),
+		rp.WithPKCE(cookieHandler),
+	}
+	relyingParty, err := rp.NewRelyingPartyOIDC(
+		context.Background(),
+		string(provider.Issuer), string(provider.ClientID), string(provider.ClientSecret),
+		provider.RedirectURI, provider.Scopes, options...,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &oIDCEndpoint{
+		relyingParty:    relyingParty,
+		jwt:             jwt,
+		tokenManagement: tokenManagement{jwt: jwt},
+		slugID:          provider.SlugID,
+		urlParams:       provider.URLParams,
+	}, nil
+}
+
+func (e *oIDCEndpoint) CollectRoutes(g *route.Group) {
+	g.GET(fmt.Sprintf("/%s/%s", e.String(), loginPath), e.buildAuthHandler(), true)
+	g.GET(fmt.Sprintf("/%s/%s", e.String(), callbackPath), e.buildCodeExchangeHandler(), true)
+	g.GET(fmt.Sprintf("/%s/%s", e.String(), logoutPath), e.logout, true)
+}
+
+func (e *oIDCEndpoint) String() string {
+	return strings.Join([]string{"oidc", e.slugID}, "-")
```

Comment:
I think I would prefer to have the route `/auth/providers/oauth/github` instead of `/auth/providers/oauth-github`.

Probably it can removed the usage of the method `String()` and the needs of having a new interface `authEndpoint` since you can define the route like that:

```go
func (e *oIDCEndpoint) CollectRoutes(g *route.Group) {
	oauthGroup = g.Group("/oauth")
	oauthGroup.GET(fmt.Sprintf("/%s", e.slugID, loginPath), e.buildAuthHandler(), false)
	oauthGroup.GET(fmt.Sprintf("/%s", e.slugID, callbackPath), e.buildCodeExchangeHandler(), true)
	oauthGroup.GET(fmt.Sprintf("/%s", e.slugID, logoutPath), e.logout, true)
```

Reply:
I simplified a lot. Iremoved the provider claim from the token and I removed the logout by provider. Let's do this like that for now and we'll see in the future if want to put some mechanism to ensure people are also logged out from providers